src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/PostgresAccessPolicyDAO.scala

```diff
@@ -932,75 +980,103 @@ class PostgresAccessPolicyDAO(protected val writeDbRef: DbReference, protected v
+          ActionResult(
+            rs.stringOpt(part.resultName.name).map(ResourceTypeName(_)),
+            rs.stringOpt(ra.resultName.action).map(ResourceAction(_)),
+            rs.booleanOpt(pa.resultName.descendantsOnly))
+        )
+      )
+    ).list().apply().groupBy(_._1)
+  }
 
-      results.map { case (policyInfo, resultsByPolicy) =>
-        val (_, memberResults, permissionsResults) = resultsByPolicy.unzip3
+  override def loadPolicyMembership(policyId: FullyQualifiedPolicyId, samRequestContext: SamRequestContext): IO[Option[AccessPolicyMembership]] = {
+    listPoliciesMemberships(policyId.resource, Option(policyId.accessPolicyName), samRequestContext).map(_.headOption.map(_.membership))
+  }
 
-        val policyMembers = unmarshalPolicyMembers(memberResults)
+  override def listAccessPolicyMemberships(resource: FullyQualifiedResourceId, samRequestContext: SamRequestContext): IO[LazyList[AccessPolicyWithMembership]] = {
+    listPoliciesMemberships(resource, None, samRequestContext)
+  }
+
+  // a variant of listPolicies but returns a structure that contains member emails and nested policy information instead of subject ids
+  private def listPoliciesMemberships(resource: FullyQualifiedResourceId, limitOnePolicy: Option[AccessPolicyName] = None, samRequestContext: SamRequestContext): IO[LazyList[AccessPolicyWithMembership]] = {
+    readOnlyTransaction("listPoliciesMemberships", samRequestContext)({ implicit session =>
+      // query to get policies and all associated roles and actions
+      val policyInfos = policyInfoQuery(resource, limitOnePolicy)
+
+      policyInfos.map { case (policyInfo, resultsByPolicy) =>
+        val (policyRoles, policyActions, policyDescendantPermissions) = unmarshalPolicyPermissions(resultsByPolicy.map(_._2))
 
-        val (policyRoles, policyActions, policyDescendantPermissions) = unmarshalPolicyPermissions(permissionsResults)
+        val policyId = FullyQualifiedPolicyId(resource, policyInfo.name)
+        // queries to get all members, there are 3 kinds, groups, users and policies
+        // policies have extra information, users and groups are just emails
+        val memberGroups = directMemberGroupEmailsQuery(policyId).list().apply().map(_.email)
+        val memberUsers = directMemberUserEmailsQuery(policyId).list().apply().map(_.email)
+        val memberPolicies = directMemberPolicyIdAndEmailsQuery(policyId).list().apply().toSet
 
-        AccessPolicy(FullyQualifiedPolicyId(FullyQualifiedResourceId(policyInfo.resourceTypeName, policyInfo.resourceId), policyInfo.name), policyMembers, policyInfo.email, policyRoles, policyActions, policyDescendantPermissions, policyInfo.public)
+        AccessPolicyWithMembership(policyInfo.name, AccessPolicyMembership(memberPolicies.map(_.policyEmail) ++ memberUsers ++ memberGroups, policyActions, policyRoles, Option(policyDescendantPermissions), Option(memberPolicies)), policyInfo.email)
       }.to(LazyList)
     })
   }
```

Comment:
`listPolicies` and `listPoliciesMemberships` are very similar - they differ only in how they map the `directMemberGroupEmailsQuery `, `directMemberUserEmailsQuery `, `directMemberUserEmailsQuery` results into their respective response objects. Is it worth factoring out a shared/common method here?

Reply:
yes, I have done so locally as well as fixing the 3n query problem